watchtower-plugin/src/convert.rs

```diff
@@ -258,6 +258,105 @@ impl TryFrom<serde_json::Value> for GetAppointmentParams {
     }
 }
 
+// Errors related to `getregistrationreceipt` command
+#[derive(Debug)]
+pub enum GetRegistrationReceiptError {
+    InvalidId(String),
+    InvalidFormat(String),
+}
+
+impl std::fmt::Display for GetRegistrationReceiptError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            GetRegistrationReceiptError::InvalidId(x) => write!(f, "{x}"),
+            GetRegistrationReceiptError::InvalidFormat(x) => write!(f, "{x}"),
+        }
+    }
+}
+
+// Parameters related to the `getregistrationreceipt` command
+#[derive(Debug)]
+pub struct GetRegistrationReceiptParams {
+    pub tower_id: TowerId,
+    pub subscription_start: Option<u32>,
+    pub subscription_expiry: Option<u32>,
+}
+
+impl TryFrom<serde_json::Value> for GetRegistrationReceiptParams {
+    type Error = GetRegistrationReceiptError;
+
+    fn try_from(value: serde_json::Value) -> Result<Self, Self::Error> {
+        match value {
+            serde_json::Value::Array(a) => {
+                let param_count = a.len();
+                if param_count == 2{
+                    Err(GetRegistrationReceiptError::InvalidFormat(("Both ends of boundary (subscription_start and subscription_expiry) are required.").to_string()))
+                } else if param_count != 1 && param_count != 3 {
+                    Err(GetRegistrationReceiptError::InvalidFormat(format!(
+                        "Unexpected request format. The request needs 1 or 3 parameter. Received: {param_count}"
+                    )))
+                } else{
+                    let tower_id = if let Some(s) = a.get(0).unwrap().as_str() {
+                        TowerId::from_str(s).map_err(|_| {
+                        GetRegistrationReceiptError::InvalidId("Invalid tower id".to_owned())
+                        })
+                    } else {
+                        Err(GetRegistrationReceiptError::InvalidId(
+                        "tower_id must be a hex encoded string".to_owned(),
+                        ))
+                    }?;
+
+                    let (subscription_start, subscription_expiry) = if let (Some(start), Some(expire)) = (a.get(1).and_then(|v| v.as_i64()), a.get(2).and_then(|v| v.as_i64())) {
+                        if start >= 0 && expire > start {
+                            (Some(start as u32), Some(expire as u32))
+                        } else {
+                            return Err(GetRegistrationReceiptError::InvalidFormat(
+                                    "Subscription_start must be a positive integer and subscription_expire must be a positive integer greater than subscription_start".to_owned(),
+                                    ));
+                        }
+                    } else if a.get(1).is_some() || a.get(2).is_some() {
```

Comment:
I don't think this branch is really necessary, is it?

We've already made sure that both the start and the end of the boundary have been provided, or none of them have. At this point is either the boundary is correct (start>=0 and end>start) or it is not, so it cannot be the case that either of them but not both are `Some(x)`

Reply:
Actually the method was also allowing text(alphabetic letters) as `subscription_start` and `subscription_expiry`. I added this branch to handle this case. Should I remove this ?

Reply:
Oh, right, that's because elements are passed as json `Values` and they can be anything really. I think this should be re-arranged then for clarity, since I didn't really get what was the point of having that else.

Can you make it so you capture the values in your `if let` as you are atm, but instead of convert them straight away you leave them as `Value`s, and then you try to convert them to `i64` or fail with a similar error to this one inside that same context?

e.g.:

```
let (subscription_start, subscription_expiry) = if let (Some(start), Some(expire)) = (a.get(1)), a.get(2)) {
    /// Try to convert them to i64, fail if they are not convertable
}
```